valueset.go

```diff
@@ -0,0 +1,261 @@
+				if currNodesetID != "" && nodesetId != currNodesetID {
+					continue
+				}
+				if err := parseRow(r); err != nil {
+					return nil, err
+				}
+			}
+		}
+		if !processed {
+			break
+		}
+	}
+	return ret, nil
+}
+
+// func buildNodesetNodes(nodeset Nodeset) []*lpg.Node {
+// 	grph := ls.NewDocumentGraph()
+// 	nodes := make([]*lpg.Node, 0)
+// 	for _, nsD := range nodeset.Data {
+// 		nodes = append(nodes, grph.NewNode(nsD.Labels, nsD.Properties))
+// 	}
+// 	return nodes
+// }
+
+// oldNodeset is nodeset pulled from DB
+func diff(oldNodeset, newNodeset Nodeset) (insertions []NodesetData, deletions []string, updates []NodesetData) {
+	// compare roots
+	if !lpg.NewStringSet(oldNodeset.Labels...).IsEqual(lpg.NewStringSet(newNodeset.Labels...)) {
+		// update
+		newNodeset.Labels = findLabelDiff(lpg.NewStringSet(oldNodeset.Labels...), (lpg.NewStringSet(newNodeset.Labels...)))
+	}
+	for oid, oldNsData := range oldNodeset.Data {
+		newNsData, exists := newNodeset.Data[oid]
+		if exists {
+			update := NodesetData{
+				ID:         oid,
+				Labels:     make([]string, 0),
+				Properties: make(map[string]interface{}),
+			}
+			if !lpg.NewStringSet(oldNsData.Labels...).IsEqual(lpg.NewStringSet(newNsData.Labels...)) {
+				// update - setting labels to those in newNodeset but not in oldNodeset
+				update.Labels = findLabelDiff(lpg.NewStringSet(newNsData.Labels...), (lpg.NewStringSet(oldNsData.Labels...)))
+			}
+			// if props are not equal, update to use newNodeset properties
+			if !reflect.DeepEqual(oldNsData.Properties, newNsData.Properties) {
+				for k, v := range newNsData.Properties {
+					update.Properties[k] = v
+				}
+			}
+			updates = append(updates, update)
+		} else {
+			deletions = append(deletions, oid)
+		}
+	}
+	for nid, newNsData := range newNodeset.Data {
+		_, exists := oldNodeset.Data[nid]
+		if !exists {
+			insertions = append(insertions, newNsData)
+		}
+	}
+	return insertions, deletions, updates
+}
+
+func LoadNodeset(tx neo4j.Transaction, nodesetId string) (Nodeset, error) {
+	query := "MATCH (root:`NODESET` {nodeset_id: $id})-[:$id]->(m) return root,m"
+	idrec, err := tx.Run(query, map[string]interface{}{"id": nodesetId})
+	if err != nil {
+		return Nodeset{}, err
+	}
+	ns := Nodeset{
+		ID:     nodesetId,
+		Labels: make([]string, 0),
+		Data:   make(map[string]NodesetData),
+	}
+	first := false
+	for idrec.Next() {
+		rec := idrec.Record()
+		if !first {
+			nsCenter := rec.Values[0].(neo4j.Node)
+			ns.Labels = nsCenter.Labels
+			first = true
+		}
+		nsLeaf := rec.Values[1].(neo4j.Node)
+		var nsLeafEntityId string
+		for k := range nsLeaf.Props {
+			if k == ls.EntityIDTerm {
+				nsLeafEntityId = k
+			}
+		}
+		ns.Data[nsLeafEntityId] = NodesetData{
+			ID:         nsLeafEntityId,
+			Labels:     nsLeaf.Labels,
+			Properties: nsLeaf.Props,
+		}
+	}
+	return ns, nil
+}
+
+func (ns Nodeset) Execute(tx neo4j.Transaction, inserts, updates []NodesetData, deletions []string) error {
```

Comment:
Don't make it a method. Get the oldNS, and new NS as args.

You need to pass in the root flag as well, so you update/delete/insert the root